### ISSUE TITLE
False positive: "None" not callable

### ISSUE THREAD

**Issue by clo-vis** [python/mypy#13631]
**To Reproduce**

```python
from typing import Callable


def my_function1(test: Callable[[int], bool] | None = None) -> None:
    if test is None:
        test = lambda value: value > 0
    test(1)


def my_function2(test: Callable[[int], bool] | None = None) -> None:
    if test is None:
        def test(value: int) -> bool:
            return value > 0
    test(2)
```

**Expected Behavior**

No errors from mypy.

**Actual Behavior**

mypy pinpoints a problem at line _test(2)_ in my_function2: "None" not callable, even though test cannot be None at that point.

Assigning a lambda expression to test (as done in my_function1) is not an alternative according to PEP 8 (says pylint)

**Your Environment**

- Mypy version used: 0.971
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used:  3.10.4

**Comment by AlexWaygood** [comment]
Already fixed on mypy master (by @hauntsaninja in #13508, I think)! Not sure whether it'll be included in mypy 0.980 or whether you'll have to wait for mypy 0.990, but a release with the fix will be out soon.
